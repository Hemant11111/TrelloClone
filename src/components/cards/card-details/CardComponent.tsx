import { useCallback, useState } from "react";
import { useDispatch } from "react-redux";

import "./CardComponent.scss";
import CardModel from "../../../model/Card";
import ListTaskComponent from "../../task/list-task/ListTaskComponent";
import InputForm from "../../../shared-components/input-form/InputForm";
import { taskMovedToAnotherCardThunk } from "../../../redux/thunk/task.thunk";
import { deleteCardThunk, editCardThunk } from "../../../redux/thunk/card.thunk";

interface CardComponentProps {
    card: CardModel;
}

export default function CardComponent(
    {
        card: {id, title, updating},
        card
    }: CardComponentProps) {

    const dispatch = useDispatch();

    const [_newCardTitle, setNewCardTitle] = useState(title);

    const handleCardDelete = useCallback(() => {
        if (window.confirm("Are you sure, you wants to delete this Card?")) {
            dispatch(deleteCardThunk(id));
        }
    }, []);

    function _udpateCard(card: CardModel) {
        dispatch(editCardThunk(card));
    }

    const handleDoubleClick = useCallback(() => {
        _udpateCard({...card, updating: true});
    }, []);


    const handleTitleChange = useCallback((e: any) => {
        setNewCardTitle(e.target.value)
    }, [setNewCardTitle]);

    const handleTitleUpdate = useCallback((e: any) => {
        _udpateCard({...card, title: _newCardTitle, updating: false});
        e.preventDefault();
    }, [_newCardTitle]);


    const handleDrop = useCallback((event: any) => {
        event.preventDefault();
        const taskId = event.dataTransfer.getData("taskId");
        dispatch(taskMovedToAnotherCardThunk(taskId, id));
    }, []);

    const handleDragOver = useCallback((event: any) => {
        event.preventDefault();
    }, []);


    return (
        <div className="card card-container" onDrop={handleDrop} onDragOver={handleDragOver}>
            {updating ?
                (
                    <InputForm
                        placeholder="Add card title..."
                        onChange={handleTitleChange}
                        onSubmit={handleTitleUpdate}
                        value={_newCardTitle}
                    />
                ) : (
                    <span className="card-title" onDoubleClick={handleDoubleClick}>{title}</span>
                )
            }

            <div className="card-tasks-container">
                <ListTaskComponent cardId={id}/>
            </div>

            <button data-testid="btn-card-delete" className="card-dlt-button" onClick={handleCardDelete}>
                <i className="fa fa-remove"/> Delete
            </button>
        </div>
    )
}