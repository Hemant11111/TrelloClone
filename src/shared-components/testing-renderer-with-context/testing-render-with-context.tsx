import { ReactElement } from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";

import store from "../../redux/store/store";

export function testingRenderWithContext(element: ReactElement) {
    return render(
        <Provider store={store}>
            {element}
        </Provider>
    )
}