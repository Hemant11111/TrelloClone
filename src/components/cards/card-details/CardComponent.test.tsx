import React from 'react';
import { screen } from "@testing-library/react";

import CardComponent from "./CardComponent";
import CardModel from "../../../model/Card";
import { testingRenderWithContext } from "../../../shared-components/testing-renderer-with-context/testing-render-with-context";


describe('Card Component', () => {
    it('should match snapshot', () => {
        const {asFragment} = testingRenderWithContext(<CardComponent card={CardModel.newInstance("test")}/>);
        expect(asFragment()).toMatchSnapshot();
    })

    it('should render delete button', () => {
        testingRenderWithContext(<CardComponent card={CardModel.newInstance("test")}/>);

        const deleteButton = screen.getByTestId("btn-card-delete");

        expect(deleteButton).toBeInTheDocument();
    })


})