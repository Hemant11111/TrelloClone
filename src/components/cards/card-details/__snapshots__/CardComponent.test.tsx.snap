// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="card card-container"
  >
    <span
      class="card-title"
    >
      test
    </span>
    <div
      class="card-tasks-container"
    >
      <div
        class="task-list-container"
      >
        <div
          class="add-task-container"
        >
          <form
            class="custom-form"
          >
            <input
              class=" input-form"
              placeholder="Add Task..."
              value=""
            />
          </form>
          <button
            class="add-task-button trello-icon-button"
            data-testid="add-task-button"
            disabled=""
          >
            <i
              class="fa fa-plus"
            />
          </button>
        </div>
      </div>
    </div>
    <button
      class="card-dlt-button"
      data-testid="btn-card-delete"
    >
      <i
        class="fa fa-remove"
      />
       Delete
    </button>
  </div>
</DocumentFragment>
`;
